typed-racket-lib/typed-racket/rep/rep-utils.rkt

```diff
@@ -24,159 +25,245 @@
 
 
 (provide (all-defined-out)
+         get-uid-count
          (for-syntax var-name))
 
+
+;; Contract and Hash related helpers
+
 (provide-for-cond-contract length>=/c)
 
 (define-for-cond-contract ((length>=/c len) l)
   (and (list? l)
        (>= (length l) len)))
 
-;; seq: interning-generated serial number used to compare Reps (type<).
-;; free-vars: cached free type variables
-;; free-idxs: cached free dot sequence variables
-;; stx: originating syntax for error-reporting
-(struct Rep (seq free-vars free-idxs) #:transparent
-  #:methods gen:equal+hash
-  [(define (equal-proc x y recur)
-     (unsafe-fx= (Rep-seq x) (Rep-seq y)))
-   (define (hash-proc x recur) (Rep-seq x))
-   (define (hash2-proc x recur) (Rep-seq x))])
-
-(define (Rep<? x y)
-  (unsafe-fx< (Rep-seq x) (Rep-seq y)))
-
-;; prop:get-values
-(define-values (prop:Rep-name Rep-name)
-  (let-values ([(prop _ accessor) (make-struct-type-property 'named)])
-    (values prop accessor)))
-
-
-;; prop:get-values
-(define-values (prop:values-fun values-fun)
-  (let-values ([(prop _ accessor) (make-struct-type-property 'values)])
-    (values prop accessor)))
-
-;; Rep-values
-(define (Rep-values x)
-  ((values-fun x) x))
+(define-for-cond-contract name-ref/c
+  (or/c identifier?
+        (cons/c exact-integer? exact-integer?)))
 
-;; prop:get-constructor
-(define-values (prop:constructor-fun Rep-constructor)
-  (let-values ([(prop _ accessor) (make-struct-type-property 'constructor)])
-    (values prop accessor)))
+(define (name-ref=? x y)
+  (cond
+    [(identifier? x)
+     (and (identifier? y) (free-identifier=? x y))]
+    [else (equal? x y)]))
 
-;; structural type info for simple/straightforward types
-;; (i.e. we store the list of field variances)
-(define-values (prop:structural structural? Type-variances)
-  (make-struct-type-property 'structural))
+(define id-table (make-free-id-table))
+(define (normalize-id id)
```

Comment:
I guess we'll have to deal, but we could consider an approach where we made the table weak.